lapce-data/src/data.rs

```diff
@@ -133,7 +136,71 @@ impl LapceData {
         let latest_release = Arc::new(None);
 
         let dirs: Vec<&PathBuf> = paths.iter().filter(|p| p.is_dir()).collect();
-        let files: Vec<&PathBuf> = paths.iter().filter(|p| p.is_file()).collect();
+
+        // Storage for parsed file, line and column.
+        let mut files: Vec<PathBuf> = vec![];
+        let mut lines: Vec<Line> = vec![];
+        let mut columns: Vec<LineCol> = vec![];
+
+        // extract file, line and column from the path
+        paths.iter().for_each(|p| {
+            let string = p
+                .file_name()
+                .expect("Failed to parse filename")
+                .to_os_string()
+                .into_string()
+                .expect("");
+            let split: Vec<_> = string.split(':').collect();
+            let (file, line, column) = match split.len() {
+                // path example: "file.txt"
+                1 => (
+                    PathBuf::from(split[0].to_string()),
+                    Line(1),
+                    LineCol { line: 1, column: 1 },
+                ),
+                // path example: "file.txt:42"
+                2 => (
+                    PathBuf::from(split[0].to_string()),
+                    Line(
+                        split[1]
+                            .parse()
+                            .expect("failed to parse line component of path"),
```

Comment:
Correct me if I'm wrong, but it seems to me that the program will crash if you specify the wrong path. Is it really necessary? Maybe it's better to handle the error and return to the default path?

Reply:
Without Colum and line

Reply:
`expect()` makes program panic

Reply:
`parse()` gives `Result<F, <F as FromStr>::Err>`. The docs for `parse()` say:
> Will return [Err](https://doc.rust-lang.org/std/str/trait.FromStr.html#associatedtype.Err) if it’s not possible to parse this string slice into the desired type.

So it's not the sequence of characters that triggers error, but rather error is caused when you have the wrong data type as input (like inserting u32 instead of PathBuf would be an example I guess).

Reply:
yes and `expect()` will make lapce panic

Reply:
Line is usize (https://github.com/lapce/lapce/blob/master/lapce-data/src/editor.rs#L113). So as I can this function  panic in case of this path: "C:\some_path" since you will try convert "some_path" to usize

Reply:
I fixed it. Is it good now?

Reply:
No :-). I think your parse function should look like this. Please write tests for it, since I wrote it myself but did not test it.

```rust
use core::fmt;

fn parse_file_line_column(
    path: &Path,
) -> Option<(PathBuf, Option<Line>, Option<LineCol>)> {
    // Trying to convert path to str and then to `reversed iterator`
    if let Some(str) = path.to_str() {
        let mut splits = str.rsplit(':');
        if let Some(first_rhs) = splits.next() {
            // Try to parse first_rhs (first right hand side element) as usize
            match first_rhs.parse::<usize>() {
                Ok(first_rhs_num) => {
                    // So first_rhs is an usize, so check the second one
                    if let Some(second_rhs) = splits.next() {
                        match second_rhs.parse::<usize>() {
                            Ok(second_rhs_num) => {
                                // The second right hand side element is numeric,
                                // so let's firstly check if the whole path without last two element
                                // is a file (remember about reversed iterator)
                                let mut str = String::new();
                                write_text_with_sep_to(splits.rev(), &mut str, ":")
                                    .ok()?;
                                // NOTE: The last element is ":", and its ok, because even if we use &[&str] we need
                                // to check the length of a slice on each iteration
                                let left_path = PathBuf::from(&str[..str.len() - 1]);
                                if left_path.is_file() {
                                    return Some((
                                        left_path,
                                        Some(Line(first_rhs_num)),
                                        Some(LineCol {
                                            line: first_rhs_num,
                                            column: second_rhs_num,
                                        }),
                                    ));
                                }
                                // It seems that even if we have some second right hand number,
                                // this does not mean that the remaining path is a file, then
                                // let's check if it changed if we add `second_rhs`?
                                // Last char of `str` is ":", so we neen to push only `second_rhs`
                                str.push_str(second_rhs);
                                let left_path = PathBuf::from(&str);
                                if left_path.is_file() {
                                    return Some((
                                        left_path,
                                        Some(Line(first_rhs_num)),
                                        None,
                                    ));
                                } else if path.is_file() {
                                    // Ok, that's not a file either. Then just check whole path
                                    return Some((path.to_path_buf(), None, None));
                                }
                            }
                            Err(_) => {
                                // The second right hand side element is not numeric,
                                // so let's firstly check if the whole path without last element
                                // is a file (remember about reversed iterator)
                                let mut str = String::new();
                                write_text_with_sep_to(splits.rev(), &mut str, ":")
                                    .ok()?;
                                // Last char of `str` is ":", so we neen to push only `second_rhs`
                                str.push_str(second_rhs);
                                let left_path = PathBuf::from(str);
                                if left_path.is_file() {
                                    return Some((
                                        left_path,
                                        Some(Line(first_rhs_num)),
                                        None,
                                    ));
                                } else if path.is_file() {
                                    // It seems that even if we have some number on the right,
                                    // this does not mean that the remaining path is a file, then
                                    // let's check if the entire path is a file?
                                    return Some((path.to_path_buf(), None, None));
                                }
                            }
                        }
                    } else if path.is_file() {
                        // It looks like we don't have a second element, so let's check if the whole
                        // path is a file? (by the way, this is most likely impossible, because it
                        // means that the entire path to the file is numeric)
                        return Some((path.to_path_buf(), None, None));
                    }
                }
                Err(_) => {
                    // Looks like the first right hand side element is not a usize,
                    // but may be the whole path is a file?
                    if path.is_file() {
                        return Some((path.to_path_buf(), None, None));
                    }
                }
            }
        }
    } else if path.is_file() {
        // It can not be converted into utf-8, but may be it is a file?
        return Some((path.to_path_buf(), None, None));
    }
    None
}

// FIXME: Unfortunately the last element will be ":", we need to think about how to handle
// it without having to allocate an unnecessary vector
fn write_text_with_sep_to<I, T, Buf>(
    mut iter: I,
    buf: &mut Buf,
    sep: T,
) -> fmt::Result
where
    Buf: fmt::Write,
    I: Iterator<Item = T>,
    T: AsRef<str>,
{
    if let Some(str) = iter.next() {
        buf.write_str(str.as_ref())?;
        buf.write_str(sep.as_ref())?;
        // call ourselves recursively
        write_text_with_sep_to(iter, buf, sep)?
    }
    fmt::Result::Ok(())
}
```

Reply:
https://github.com/panekj/lapce/commit/e9e091f34ee48440145f5b8e76662b9666eeebca

Reply:
I'm a bit suspicious on the idea that `None` should be changed into something else. What does `scroll_offset` mean? It requires an input of `Some<Ve2<f64, f64>>`. It sounds like this is about where on the screen you want the destination line to show on. 

The second `None`, namely `History`, I'm not sure why history would be relevant here. You're just opening a new file from scratch and want it to jump to a line. There is no history-dependence in this operation.

Reply:
I changed the code, using your suggested code as an inspiration source. Let me know if it's good now. @panekj  says we shouldn't have tests and @JustForFun88  says we should have tests. I'll let you two discuss with each other and come to some agreement. Until then, I'm ready for new feedback.